File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/adapter/SkillTasksRecyclerViewAdapter.java
package com.habitrpg.android.habitica.ui.adapter;

import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.models.tasks.Task;

import java.util.UUID;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.OrderedRealmCollection;
import io.realm.RealmRecyclerViewAdapter;
import rx.Observable;
import rx.subjects.PublishSubject;


public class SkillTasksRecyclerViewAdapter extends RealmRecyclerViewAdapter<Task, SkillTasksRecyclerViewAdapter.ViewHolder> {

    private PublishSubject<Task> taskSelectionEvents = PublishSubject.create();

    public SkillTasksRecyclerViewAdapter(@Nullable OrderedRealmCollection<Task> data, boolean autoUpdate) {
        super(data, autoUpdate);
    }

    @Override
    public long getItemId(int position) {
        if (getData() != null) {
            Task task = getData().get(position);
            if (task.getId() != null && task.getId().length() == 36) {
                return UUID.fromString(task.getId()).getMostSignificantBits();
            }
        }
        return UUID.randomUUID().getMostSignificantBits();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.skill_task_item_card, parent, false);
        return new SkillTasksRecyclerViewAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        if (getData() != null) {
            holder.bindHolder(getData().get(position));
        }
    }

    public Observable<Task> getTaskSelectionEvents() {
        return taskSelectionEvents.asObservable();
    }


    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        public Task task;
        protected android.content.res.Resources resources;

        @BindView(R.id.titleTextView)
        TextView titleTextView;
        @BindView(R.id.notesTextView)
        TextView notesTextView;
        @BindView(R.id.rightBorderView)
        View rightBorderView;

        public ViewHolder(View itemView) {
            super(itemView);

            itemView.setOnClickListener(this);
            itemView.setClickable(true);

            ButterKnife.bind(this, itemView);

            resources = itemView.getResources();
        }

        void bindHolder(Task task) {
            this.task = task;
            titleTextView.setText(task.getText());
            if (task.getNotes() == null || task.getNotes().length() == 0) {
                notesTextView.setVisibility(View.GONE);
            } else {
                notesTextView.setVisibility(View.VISIBLE);
                notesTextView.setText(task.getNotes());
            }
            rightBorderView.setBackgroundResource(task.getLightTaskColor());
        }

        @Override
        public void onClick(View v) {
            if (v.equals(itemView)) {
                taskSelectionEvents.onNext(task);
            }
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/adapter/SkillTasksRecyclerViewAdapter.kt
package com.habitrpg.android.habitica.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.databinding.SkillTaskItemCardBinding
import com.habitrpg.android.habitica.models.tasks.Task
import java.util.UUID

class SkillTasksRecyclerViewAdapter :
    BaseRecyclerViewAdapter<Task, SkillTasksRecyclerViewAdapter.TaskViewHolder>() {
    var onTaskSelection: ((Task) -> Unit)? = null

    override fun getItemId(position: Int): Long {
        val task = getItem(position)
        if (task?.id?.length == 36) {
            return UUID.fromString(task.id).mostSignificantBits
        }
        return UUID.randomUUID().mostSignificantBits
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): TaskViewHolder {
        val view =
            LayoutInflater.from(parent.context)
                .inflate(R.layout.skill_task_item_card, parent, false)
        return TaskViewHolder(view)
    }

    override fun onBindViewHolder(
        holder: TaskViewHolder,
        position: Int,
    ) {
        holder.bindHolder(data[position])
    }

    inner class TaskViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        private val binding = SkillTaskItemCardBinding.bind(itemView)
        var task: Task? = null

        init {
            itemView.setOnClickListener(this)
            itemView.isClickable = true
        }

        internal fun bindHolder(task: Task) {
            this.task = task
            binding.titleTextView.text = task.markdownText { binding.titleTextView.text = it }
            if (task.notes?.isEmpty() == true) {
                binding.notesTextView.visibility = View.GONE
            } else {
                binding.notesTextView.visibility = View.VISIBLE
                binding.notesTextView.text = task.markdownNotes { binding.notesTextView.text = it }
            }
            binding.rightBorderView.setBackgroundColor(
                ContextCompat.getColor(
                    itemView.context,
                    task.lightTaskColor,
                ),
            )
        }

        override fun onClick(v: View) {
            if (v == itemView) {
                task?.let {
                    onTaskSelection?.invoke(it)
                }
            }
        }
    }
}
